### PR TITLE
Fix beatmap carousel not maintaining selection if currently selected beatmap is updated

### LINKED ISSUE #19328: Updating a selected beatmap may not follow selection at song select
### Discussed in https://github.com/ppy/osu/discussions/19327

<div type='discussions-op-text'>

<sup>Originally posted by **Flamiii** July 23, 2022</sup>
Updating a map can cause it to seemingly disappear as it gets moved from its original position in the beatmap carousel to another location. For example, if you're sorting by difficulty and are updating a 6.8* map that was edited and is now 7.2*, the map will disappear from view. I think it would be better behavior if the game brought the user to the newly updated map.

https://user-images.githubusercontent.com/74088315/180602283-144923e6-3c4a-4528-a352-69469af2b804.mp4

</div>

### PR DESCRIPTION
**Pull request by peppy**
RFC. I've tested this to work, but adding tests is a bit involved. Want to make sure the solution seems okay before adding them.

- Closes #19328
- [x] Add tests

### CODE REVIEW

**Comment by LittleEndu** [comment]
Looking at the code, doesn't this fail when the update to the beatmap is a deletion of the selected difficulty?

**Comment by peppy** [comment]
Can you explain what you mean and which line is at fault? I can't read minds.

File: osu.Game/Screens/Select/BeatmapCarousel.cs [hunk 1]
```diff
@@ -265,6 +265,38 @@ private void beatmapSetsChanged(IRealmCollection<BeatmapSetInfo> sender, ChangeS
...
+                    foreach (int i in changes.NewModifiedIndices)
+                    {
+                        var beatmapSetInfo = sender[i];
+
+                        foreach (var beatmapInfo in beatmapSetInfo.Beatmaps)
+                        {
+                            // Best effort matching. We can't use ID because in the update flow a new version will get its own GUID.
+                            bool selectionMatches =
+                                ((IBeatmapMetadataInfo)beatmapInfo.Metadata).Equals(SelectedBeatmapInfo.Metadata)
+                                && beatmapInfo.DifficultyName == SelectedBeatmapInfo.DifficultyName;
```
Comment (LittleEndu) [inline osu.Game/Screens/Select/BeatmapCarousel.cs:289]: I meant this. Isn't this true only if the difficulty also exists in the updated beatmap.

I would think if the beatmap set's metadata matches, it's OK to select any difficulty, not necessarily the one that was previously selected.

Comment (peppy) [reply osu.Game/Screens/Select/BeatmapCarousel.cs]: Right. A `BeatmapSet`'s metadata is currently just a reference to the first beatmap in the set, so it gets a bit difficult to guarantee anything there.

There are other scenarios such as if the difficulty name is changed (or the title of the beatmap is changed) which will lose selection in a similar way, so if we want to be more inclusive then a fallback selection can be added. See https://github.com/ppy/osu/pull/19409/commits/24d75612e222fe3e055d1f5fabf623727dc81727.

**Comment by LittleEndu** [comment]
I want to try this. Should I wait for tests to be added or can I test in prod?

**Comment by frenzibyte** [comment]
You should be able to test this in-game by modifying the beatmap however you want then update and see if it selects properly.

**Comment by peppy** [comment]
To clarify, you can use the staging/dev environment and download a beatmap, then make a change via realm studio or the editor.

File: osu.Game/Screens/Select/BeatmapCarousel.cs [hunk 2]
```diff
@@ -265,6 +265,41 @@ private void beatmapSetsChanged(IRealmCollection<BeatmapSetInfo> sender, ChangeS
 
             foreach (int i in changes.InsertedIndices)
                 UpdateBeatmapSet(sender[i].Detach());
+
+            if (changes.DeletedIndices.Length > 0)
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Screens/Select/BeatmapCarousel.cs:269]: That's not good for the methodology used in this PR. I've asked about this over at realm so we can better understand the scenario (https://github.com/realm/realm-dotnet/discussions/2634#discussioncomment-3277702).

[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Screens/Select/BeatmapCarousel.cs:269]: @frenzibyte I've tried and can't reproduce what you're seeing. Can you share your process for getting into that state? Are you making an edit in the editor or something else?

[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Screens/Select/BeatmapCarousel.cs:269]: Managed to reproduce this, and fixed the case.

**Comment by peppy** [comment]
This was incorrectly blocked, should actually be okay to review.

**Review by frenzibyte — DISMISSED** [review]

**Comment by frenzibyte** [comment]
@peppy Do you still plan to add tests to this PR? It's blocking merge currently.

**Comment by peppy** [comment]
will see what i can do

**Comment by peppy** [comment]
I've added test coverage for this (confirmed to fail on `master`).

**Review by frenzibyte — APPROVED** [review]

**Pull request merged**